plugins/CoreUpdater/Updater.php

```diff
@@ -112,19 +117,58 @@ public function updatePiwik($https = true)
             $this->verifyDecompressedArchive($extractedArchiveDirectory);
             $messages[] = $this->translator->translate('CoreUpdater_VerifyingUnpackedFiles');
 
-            $disabledPluginNames = $this->disableIncompatiblePlugins($newVersion);
-            if (!empty($disabledPluginNames)) {
-                $messages[] = $this->translator->translate('CoreUpdater_DisablingIncompatiblePlugins', implode(', ', $disabledPluginNames));
-            }
+            // we need to load the marketplace already here, otherwise it will use the new, updated file in Piwik 3
+            $marketplace = new MarketplaceApiClient();
+            require_once PIWIK_DOCUMENT_ROOT . '/plugins/CorePluginsAdmin/PluginInstaller.php';
+            require_once PIWIK_DOCUMENT_ROOT . '/plugins/CorePluginsAdmin/MarketplaceApiException.php';
 
             $this->installNewFiles($extractedArchiveDirectory);
             $messages[] = $this->translator->translate('CoreUpdater_InstallingTheLatestVersion');
+
         } catch (ArchiveDownloadException $e) {
             throw $e;
         } catch (Exception $e) {
             throw new UpdaterException($e, $messages);
         }
 
+        try {
+            if (CorePluginsAdmin::isMarketplaceEnabled()) {
+                $messages[] = $this->translator->translate('CoreUpdater_CheckingForPluginUpdates');
+
+                $pluginManager = \Piwik\Plugin\Manager::getInstance();
+                $pluginManager->loadAllPluginsAndGetTheirInfo();
+                $loadedPlugins = $pluginManager->getLoadedPlugins();
+
+                MarketplaceApiClient::clearAllCacheEntries();
+                StaticContainer::getContainer()->set('marketplacePiwikVersion', $newVersion);
+
+                $pluginsWithUpdate = $marketplace->checkUpdates($loadedPlugins);
+
+                foreach ($pluginsWithUpdate as $pluginWithUpdate) {
+                    $pluginName = $pluginWithUpdate['name'];
+
+                    $messages[] = $this->translator->translate('CoreUpdater_UpdatingPluginXToVersionY',
+                                                               array($pluginName, $pluginWithUpdate['version']));
+
+                    $pluginInstaller = new PluginInstaller($pluginName);
+                    $pluginInstaller->installOrUpdatePluginFromMarketplace();
+                }
+            }
+        } catch (MarketplaceApiException $e) {
+            // there is a problem with the connection to the server, ignore for now
+        } catch (Exception $e) {
+            throw new UpdaterException($e, $messages);
+        }
+
+        try {
+            $disabledPluginNames = $this->disableIncompatiblePlugins($newVersion);
```

Comment:
If something goes wrong with the network when updating the plugin via the Marketplace above, we disable the plugin after all